gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/event/EventProducerFactory.java

```diff
@@ -0,0 +1,81 @@
+package org.apache.gobblin.event;
+
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigFactory;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.gobblin.broker.EmptyKey;
+import org.apache.gobblin.broker.ResourceInstance;
+import org.apache.gobblin.broker.gobblin_scopes.GobblinScopeTypes;
+import org.apache.gobblin.broker.iface.ConfigView;
+import org.apache.gobblin.broker.iface.NoSuchScopeException;
+import org.apache.gobblin.broker.iface.NotConfiguredException;
+import org.apache.gobblin.broker.iface.ScopeType;
+import org.apache.gobblin.broker.iface.ScopedConfigView;
+import org.apache.gobblin.broker.iface.SharedResourceFactory;
+import org.apache.gobblin.broker.iface.SharedResourceFactoryResponse;
+import org.apache.gobblin.broker.iface.SharedResourcesBroker;
+import org.apache.gobblin.configuration.SourceState;
+import org.apache.gobblin.configuration.State;
+import org.apache.gobblin.configuration.WorkUnitState;
+
+
+/**
+ * A factory to create shared {@link EventProducer} instance
+ */
+@Slf4j
+public class EventProducerFactory<T, S extends ScopeType<S>> implements SharedResourceFactory<EventProducer<T>, EmptyKey, S> {
+  private static final String FACTORY_NAME = "eventProducerFactory";
+  private static final String CONF_NAMESPACE = "eventProducer";
+
+  @Override
+  public String getName() {
+    return FACTORY_NAME;
+  }
+
+  @Override
+  public SharedResourceFactoryResponse<EventProducer<T>> createResource(SharedResourcesBroker<S> broker,
+      ScopedConfigView<S, EmptyKey> config)
+      throws NotConfiguredException {
+    return new ResourceInstance<>(createEventProducer(config.getConfig()));
+  }
+
+  public String getConfNamespace() {
+    return CONF_NAMESPACE;
+  }
+
+  public EventProducer<T> createEventProducer(Config config) {
+    return new EventProducer<>(config);
+  }
+
+  @Override
+  public S getAutoScope(SharedResourcesBroker<S> broker, ConfigView<S, EmptyKey> config) {
+    return broker.selfScope().getType();
```

Comment:
Why is the producer created at self-scope? This means it will not be shared among different tasks for example.

Reply:
It's hard for me to figure out what's the right auto scope as a `EventProducer` can be used just within a task or different tasks of a job. I would prefer whoever uses it specifies an explicit scope.